### PR TITLE
Have persistent DHCP leases file on VRs and cleanup /etc/hosts on VM deletion

### LINKED ISSUE #3272: VR DHCP lease file not persistent
##### ISSUE TYPE
* Bug Report - BLOCKER
* Improvement request 

##### COMPONENT NAME
* VR dnsmasq DHCP

##### CLOUDSTACK VERSION
* ACS 4.11.2
* ACS 4.12 not tested but assumed issues are still present looking at code changes
* Regressions from / previously present in ACS 4.9 / 4.10 / 4.11

##### REFERENCES
1) PR2427 - CLOUDSTACK-10252: Delete dnsmasq leases file on restart /  https://github.com/apache/cloudstack/pull/2427
2) Old Jira issue: VR DNSmasq lease file not persistent across DHCP actions
 / https://issues.apache.org/jira/browse/CLOUDSTACK-10204
3) Old Jira issue: New VR after network restart with cleanup is missing DNS and DHCP entries
 / https://issues.apache.org/jira/browse/CLOUDSTACK-10167
4) Finite lease introduction in https://github.com/apache/cloudstack/pull/1547
5) Also check related issue https://github.com/apache/cloudstack/issues/3273

##### CONFIGURATION
* ACS4.11.2
* Isolated guest networks and VPC networks

##### OS / ENVIRONMENT
* Any

##### SUMMARY

1) Leases file is not persistent across network operations:
* Refer to ref #2 above.
* Leases continues to be not persistent across dnsmasq restarts / additions - i.e. leases file is wiped and leases are not remembered.
* This causes the guest VM dhcp client to get an "NAK" DHCP deny on DHCP renewal, and have to go through the normal procedure again of DISCOVER>OFFER>REQUEST>ACK. For some guest OS versions (RedHat has shown this) where static routes are used the initial "NAK" will break these static routes, leading to problems in the guest OS, i.e. the VM will get a new lease but will have lost it's static routes.

---
2) Improvement request: Lease time for DHCP options is still hard coded to randint(700,760) hours:
https://github.com/apache/cloudstack/blob/master/systemvm/debian/opt/cloud/bin/cs/CsDhcp.py#L145
* Since this ends up in the packaged systemvm.iso file dynamically managing this might be a problem
* When looking at this issue could 5 mins be spent considering if the lease time could be put in as a global setting
---
##### STEPS TO REPRODUCE
1) Leases file is not persistent across network operations:
* Create VM
* Check contents of /var/lib/misc/dnsmasq.leases
* Create second VM
* Observe contents of /var/lib/misc/dnsmasq.leases is wiped
* Now wipe the lease file on the first VM and restart network
* Observe the /var/log/dnsmasq.log shows entries for a full  DISCOVER>OFFER>REQUEST>ACK process rather than a renew
---

##### EXPECTED RESULTS
* Leases persistent throughout lifetime of VM

##### ACTUAL RESULTS
* Leases not persistent causing static route problems on guest VMs

### LINKED ISSUE #3354: Upon VM removal, VM records not removed from /etc/hosts
##### ISSUE TYPE

 * Bug Report
 * Improvement Request
 * Enhancement Request
 * Feature Idea
 * Documentation Report
 * Other

##### COMPONENT NAME

~~~
VR
~~~

##### CLOUDSTACK VERSION

~~~
4.11.x
~~~

##### CONFIGURATION

NA

##### OS / ENVIRONMENT

NA

##### SUMMARY

When VM is deleted and expunged, VM's record is not removed from /etc/hosts file and thus DNSmasq will continue to do DNS resolution for this VM's

##### STEPS TO REPRODUCE

Destroy and expunge the VM, observe /etc/hosts of the related VR - VM's name and IP address are present.

~~~

~~~

##### EXPECTED RESULTS

After VM is expunged/deleted, it's records/entries should be properly removed from /etc/hosts (and similar, not part of this issues, from DHCP related files)
~~~

~~~

##### ACTUAL RESULTS

VM records not deleted from /etc/hosts
~~~

~~~

### PR DESCRIPTION
**Pull request by nvazquez**
## Description

Since the CloudStack virtual router was redesigned on version 4.6 it has been observed that the DHCP leases file is not persistent across network operations. This causes conflicts on guest VMs static IPs, causing these static IPs to not be renewed by the DHCP server running on isolated and VPC networks' virtual routers (dnsmasq). On stopping or destroying a VM, its dhcp/dns records are not removed from the virtual router causing ghost effects.

Fixes #3272 
Fixes #3354 

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
Tested on local environment, 1x KVM host, Isolated network
- Create VMs on isolated network
- Restart network

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by onitake** [comment]
Is the fix implemented by this PR ready to be tested?
The WIP tag is slightly confusing...
I'd like to test it on one of our affected networks, if possible.

**Comment by nvazquez** [comment]
Hi @onitake the WIP tag stands for 'Work in progress' which means this is still not ready. Having said that, you can test it now but you may expect new changes

File: systemvm/debian/opt/cloud/bin/cs/CsDhcp.py [hunk 1]
```diff
@@ -114,8 +114,23 @@ def configure_server(self):
...
+            for host in open(DHCP_HOSTS):
+                macs_dhcphosts.append(host.split(',')[0])
+
+            removed = 0
+            for leaseline in open(LEASES):
+                lease = leaseline.split(' ')
+                mac = lease[1]
+                ip = lease[2]
+                if mac not in macs_dhcphosts:
+                    cmd = "dhcp_release %s %s" % (ip, mac)
```
Comment (gildegoma) [inline systemvm/debian/opt/cloud/bin/cs/CsDhcp.py]: @nvazquez the dhcp_release `<interface>` parameter is missing here

See https://manpages.debian.org/stretch/dnsmasq-utils/dhcp_release.1.en.html

**Comment by rhtyd** [comment]
@nvazquez can you address review comments and also advise if this PR is ready for review and testing?

**Comment by rhtyd** [comment]
@nvazquez while you're on it, can you also look at another similar issue related to VRs - https://github.com/apache/cloudstack/issues/3273?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@nvazquez can you also fix pylint error:
```
************* Module cleanuphosts
E: 75,78: Instance of 'Exception' has no 'strerror' member (no-member)
-----------------------------------
Your code has been rated at 9.99/10
+'[' 2 -gt 0 ']'
+echo 'Pylint failed, please check your code'
Pylint failed, please check your code
```

File: systemvm/debian/opt/cloud/bin/cs/CsDhcp.py (hunk 1 above)
Comment (nvazquez) [reply systemvm/debian/opt/cloud/bin/cs/CsDhcp.py]: Thanks I'll fix this

Comment (rhtyd) [reply systemvm/debian/opt/cloud/bin/cs/CsDhcp.py]: Fixed @gildegoma @nvazquez

**Comment by rhtyd** [comment]
Rebased against latest 4.11, fixed pylint error and dhcp_release syntax issue.

**Comment by rhtyd** [comment]
@nvazquez fixes for #3272 and #3354 LGTM, but I found a weird issue:
- Deploy two small VMs
- Destroy one of the small VMs, verify that their entries are not seen in /etc/hosts and /etc/dhcphosts.txt
- Deploy another new small VM
- Verify that the new small VM's entry exists, but not the delete VM's entry suddenly show up as well (possibly something to do with CsDhcp.py and databags)
Please continue with the investigation and fix for the above edge-case and also fix for  #3273

**Comment by nvazquez** [comment]
@rhtyd changed the implementation, no need for a new script. Marking entries as 'remove=True' solves the issue. This solves #3272 and #3354.
Fix for #3273 is the only remaining

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Let's do #3273 in a separate PR.

**Comment by server24** [comment]
@nvazquez @rhtyid this breaks Basic Networking (#3497) whenever a VM is stopped or expunged.

**Comment by rhtyd** [comment]
Can you check and possibly send a PR @nvazquez ? Thanks.

**Comment by nvazquez** [comment]
Sure I'll take a look tonight

**Comment by GabrielBrascher** [comment]
Please ignore the message above, just saw the PR #3501 fixing it.

Thanks for the PR @nvazquez.

**Pull request merged**